### PR TITLE
Make `BorrowedFd::borrow_raw` a const fn.

### PR DESCRIPTION
**Pull request by sunfishcode**
Making `BorrowedFd::borrow_raw` a const fn allows it to be used to
create a constant `BorrowedFd<'static>` holding constants such as
`AT_FDCWD`. This will allow [`rustix::fs::cwd`] to become a const fn.

For consistency, make similar changes to `BorrowedHandle::borrow_raw`
and `BorrowedSocket::borrow_raw`.

[`rustix::fs::cwd`]: https://docs.rs/rustix/latest/rustix/fs/fn.cwd.html

r? @joshtriplett

### CODE REVIEW

**Comment by sunfishcode** [comment]
r? rust-lang/libs-api @rustbot label +T-libs-api -T-libs

**Comment by sunfishcode** [comment]
I marked this T-libs-api because it changes changes `BorrowedFd::borrow_raw`, which is currently an unstable API function, to be a `const fn`.

As context, this https://github.com/rust-lang/rust/issues/87074 is proposing to stabilize this API, though that can happen with or without this PR, since adding `const` to a function is a backwards-compatible change.

**Comment by joshtriplett** [comment]
Seems reasonable to me.

@bors r+

File: library/std/src/os/windows/io/socket.rs [hunk 1]
```diff
@@ -71,8 +71,8 @@ impl BorrowedSocket<'_> {
     /// `INVALID_SOCKET`.
     #[inline]
     #[unstable(feature = "io_safety", issue = "87074")]
-    pub unsafe fn borrow_raw(socket: RawSocket) -> Self {
-        debug_assert_ne!(socket, c::INVALID_SOCKET as RawSocket);
+    pub const unsafe fn borrow_raw(socket: RawSocket) -> Self {
+        assert!(socket != c::INVALID_SOCKET as RawSocket);
```
Comment (klensy) [inline library/std/src/os/windows/io/socket.rs:75]: was `debug_assert_ne`, but became `assert`, so will be called at not debug builds. Why this change?

Comment (sunfishcode) [reply library/std/src/os/windows/io/socket.rs:75]: This makes it consistent with the unix version of `borrowed_fd`, which does a plain `assert`. Which is there because the previous unix implementation code `FileDesc::new` used a plain `assert`.

Also, I think it makes sense; this is a new API, and we're expecting to migrate mature codebases to use it, so it makes sense to be extra careful that old and new code agree on when and where `INVALID_SOCKET` can appear. Also, `OwnedSocket` uses `rustc_layout_scalar_valid_range_*` layout tricks that assume that `INVALID_SOCKET` can't appear, so it could easily produce some dramatic UB if the invariant is violated.

Comment (klensy) [reply library/std/src/os/windows/io/socket.rs:75]: Make sense.
